pyhap/accessory_driver.py

```diff
@@ -859,118 +860,99 @@ def set_characteristics(self, chars_query, client_addr):
-            if expire_time is None or time.time() > expire_time:
-                expired = True
+            expired = expire_time is None or time.time() > expire_time
 
-        for cq in chars_query[HAP_REPR_CHARS]:
-            aid, iid = cq[HAP_REPR_AID], cq[HAP_REPR_IID]
-            setter_results.setdefault(aid, {})
+        to_update = (query for query in queries if HAP_REPR_VALUE in query or expired)
+        for query in to_update:
+            aid, iid = query[HAP_REPR_AID], query[HAP_REPR_IID]
+            value = query.get(HAP_REPR_VALUE, None)
+            write_response_requested = query.get(HAP_REPR_WRITE_RESPONSE, False)
 
-            if HAP_REPR_WRITE_RESPONSE in cq:
-                setter_responses.setdefault(aid, {})
-                had_write_response = True
+            acc = self.accessory if self.accessory.aid == aid else None
+            acc = acc or self.accessory.accessories.get(aid)
 
-            if expired:
-                setter_results[aid][iid] = HAP_SERVER_STATUS.INVALID_VALUE_IN_REQUEST
-                had_error = True
-                continue
+            char = acc.get_characteristic(aid, iid)
 
-            if HAP_PERMISSION_NOTIFY in cq:
-                char_topic = get_topic(aid, iid)
-                action = "Subscribed" if cq[HAP_PERMISSION_NOTIFY] else "Unsubscribed"
-                logger.debug(
-                    "%s client %s to topic %s", action, client_addr, char_topic
-                )
-                self.async_subscribe_client_topic(
-                    client_addr, char_topic, cq[HAP_PERMISSION_NOTIFY]
-                )
+            set_result, set_result_value = HAP_SERVER_STATUS.INVALID_VALUE_IN_REQUEST, None
+            if value is not None:
+                set_result, set_result_value = _wrap_char_setter(char, value, client_addr)
 
-            if HAP_REPR_VALUE not in cq:
-                continue
+            value_in_response = set_result_value is not None and write_response_requested
 
-            updates.setdefault(aid, {})[iid] = cq[HAP_REPR_VALUE]
+            query_results.setdefault(aid, {})[iid] = {
+                HAP_REPR_STATUS: set_result,
+                **({HAP_REPR_VALUE: set_result_value} if value_in_response else {}),
+            }
 
-        for aid, new_iid_values in updates.items():
-            if self.accessory.aid == aid:
-                acc = self.accessory
-            else:
-                acc = self.accessory.accessories.get(aid)
+            char_to_iid[char] = iid
 
-            updates_by_service = {}
-            char_to_iid = {}
-            for iid, value in new_iid_values.items():
-                # Characteristic level setter callbacks
-                char = acc.get_characteristic(aid, iid)
+            updates_by_accessories_services \
+                .setdefault(acc, {}) \
+                .setdefault(char.service, {}) \
+                .update({char: value})
```

Comment:
Please format with `black`

Also `setdefault` still creates an empty dict and throws it away.  We should avoid doing that in a loop

Reply:
Gonna look into that.